### DISCUSSION TITLE
Step 4 failing with 401 (Unauthorized)

### DISCUSSION THREAD

**Discussion by btown** [timdorr/tesla-api#563] [Q&A] [answered]
I had this working, but just a few weeks ago it stopped working.  Steps 1-3 work just fine and I successfully get a bearer token in step 3 but can't seem to exchange it for an auth token in step 4.  Here's what I'm sending:

POST https://owner-api.teslamotors.com/oauth/token
Authorization: Bearer eci(actual bearer token clipped for purposes of posting this of course)OiJSUzI1NiIs
User-Agent: AutomationAccess
Content-Type: application/json

{

    "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
    "client_id": "81527cff06843c8634fdc09e8ac0abefb46ac849f38fe1e431c2ef2106796384",
    "client_secret": "c7257eb71a564034f9419ee651c7d0e5f7aa6bfbd18bafb5c5c033b093bb2fa3"
  }

What am I doing wrong?  Thanks!

**Comment by IMgoRt** [comment, marked as answer]
That is no longer required any more
The taken from step 3 is now used for 8 hours and then you exchange that for a new token

**Reply by btown** [reply]
Thanks for that detail I was missing!!

So, we have to continue renewing within every 8 hours in order to avoid starting over, right?  Can we do that indefinitely, or is there an expiration period or limit to the number of times we can keep exchanging for a new token?

**Reply by BJReplay** [reply]
You need to renew the token before it expires to avoid starting over.

It is unknown how long you can keep doing this, but people have been doing it successfully since the new auth scheme came in around March 22.

There is some discussion that you can actually use the renewal token for auth - which might last for longer than eight hours - but it is unclear if this is the case, and, if so, how long you can use it for.  The renewal token doesn't have an expiration period, so you can't programmatically work out how long you can wait before renewing it.

The old token would expire in 45 days.  The new token will expire in 8 hours.

You should plan on renewing the token before it expires in eight hours.

**Reply by IMgoRt** [reply]
Correct, use the refresh token to create another access token - that will come with another refresh token (which so far has been identical) which you use in 28800 seconds, if you let the access expire the refresh remains valid 
When you refresh, store the time that you do that as there's no longer a created at field

**Reply by btown** [reply]
Perfect! I'm now refreshing every 7 hours and 45 minutes and everything's working great again.  Thank you both for the help!

**Reply by GeorgeIoak** [reply]
> The token from step 3 is now used for 8 hours and then you exchange that for a new token

Can anyone point me to some example code that does this, preferably in Python? I have code that works once I have the token but I'm using the Android app and storing the token in a key which my retrieval program reads and uses.

**Comment by tapwork** [comment]
Yes it seems that step 4 is not needed anymore.
I am using the bearer directly. but this token is only valid for 8 hours.  
If expired I would refresh with the refresh token (taken from step 3).
